operations/paperdl/paperdl.go

```diff
@@ -0,0 +1,157 @@
+
+import (
+	"encoding/json"
+	"errors"
+	"github.com/hashicorp/go-version"
+	"github.com/pufferpanel/pufferpanel/v3"
+	"github.com/pufferpanel/pufferpanel/v3/files"
+	"github.com/pufferpanel/pufferpanel/v3/logging"
+	"github.com/pufferpanel/pufferpanel/v3/utils"
+	"net/http"
+	"net/url"
+	"path"
+	"strings"
+)
+
+const VersionsUrl = "https://fill.papermc.io/v3/projects/paper/versions"
+const BuildUrl = "https://fill.papermc.io/v3/projects/paper/versions/${mcVersion}/builds/${build}"
+var UserAgent = pufferpanel.Display + " https://github.com/pufferpanel/pufferpanel"
+
+type PaperDl struct {
+	MinecraftVersion string
+	Build            string
+	Filename         string
+}
+
+func (op PaperDl) Run(args pufferpanel.RunOperatorArgs) pufferpanel.OperationResult {
+	env := args.Environment
+
+	if op.MinecraftVersion == "latest" {
+		logging.Info.Printf("PaperDL got Minecraft version 'latest', looking up latest version supported by Paper")
+		mcVersion, err := getLatestMCVersion()
+		if err != nil {
+			return pufferpanel.OperationResult{Error: err}
+		}
+		op.MinecraftVersion = mcVersion
+	}
+
+	dlUrl, hash, err := op.getDownloadUrlAndHash(env)
+	if err != nil {
+		return pufferpanel.OperationResult{Error: err}
+	}
+
+	localFile, err := pufferpanel.FileFromCacheOrDownload(dlUrl, hash, pufferpanel.FileHashSHA256, env)
+	if err != nil {
+		return pufferpanel.OperationResult{Error: err}
+	}
+
+
+	//copy from the cache
+	err = files.CopyFile(localFile, path.Join(env.GetRootDirectory(), op.Filename))
+	if err != nil {
+		return pufferpanel.OperationResult{Error: err}
+	}
+
+	return pufferpanel.OperationResult{Error: nil}
+}
+
+func getLatestMCVersion() (string, error) {
+	path, err := url.Parse(VersionsUrl)
+	if err != nil {
+		return "", err
+	}
+
+	request := &http.Request{
+		Method: "GET",
+		URL: path,
+		Header: http.Header{},
+	}
+	request.Header.Add("user-agent", UserAgent)
+
+	response, err := pufferpanel.Http().Do(request)
+	defer utils.CloseResponse(response)
+	if err != nil {
+		return "", err
+	}
+
+	var versions PaperVersionsResponse
+	err = json.NewDecoder(response.Body).Decode(&versions)
+	if err != nil {
+		return "", err
+	}
+
+	latest, _ := version.NewVersion("0.0")
+	for _, v := range versions.Versions {
+		if ver, err := version.NewVersion(v.VersionInfo.Id); err == nil && latest.LessThan(ver) {
+			latest = ver
+		} else {
+			logging.Info.Printf("failed to parse version '%s', %s", v, err)
+		}
+	}
+
+	logging.Info.Printf("Latest Minecraft version supported by Paper is %s", latest.Original())
+	return latest.Original(), nil
+}
+
+func (op PaperDl) getDownloadUrlAndHash(env *pufferpanel.Environment) (string, string, error) {
+	path, err := url.Parse(strings.Replace(strings.Replace(BuildUrl, "${mcVersion}", op.MinecraftVersion, -1), "${build}", op.Build, -1))
+	if err != nil {
+		return "", "", err
+	}
+
+	request := &http.Request{
+		Method: "GET",
+		URL: path,
+		Header: http.Header{},
+	}
+	request.Header.Add("user-agent", UserAgent)
```

Comment:
User-Agent is the correct casing

Reply:
Yeah, there's no standard. Even headers don't have a standard. But, User-Agent is the "standard" casing for that field.